runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+/*
+ * Copyright 2004-2018 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chplrt.h"
+#include "chpl-thread-local-storage.h"
+#include "chpl-mem.h"
+#include "chpl-atomics.h"
+#include "chpl-tasks.h"
+#include "chpl-qsbr.h"
+
+// Number of checkpoints to pass before processing
+#define CHPL_QSBR_CHECKPOINTS_PER_PROCESS 8
+
+
+
+static atomic_uint_least64_t global_epoch = 0;
+
+// Each node in the deferred list has associated with it
+// the data (and numData, useful for determining if it
+// is an array and should be treated as a 'void **') and
+// the targetEpoch which is the epoch required for all threads
+// to have observed before the data can be deleted. By making
+// use of the fact that (barring integer overflow) the largest
+// targetEpoch is always pushed as the head, we have a relative
+// sorting order. If we ask for some targetEpoch X, and for some
+// node N we have X >= N.targetEpoch, then we can claim N and
+// N.next and so and so forth since N.targetEpoch >= N.next.targetEpoch. 
+struct defer_node {
+  uint64_t targetEpoch;
+  void *data;
+  int numData;
+  struct defer_node *next;
+};
+
+// Thread-local data that keeps track of our observed global state.
+struct tls_node {
+  /*
+    The number of checkpoints passed through
+  */
+  uint64_t passedCheckpoints;
+  /*
+    The current observed epoch. This is used to ensure that writers
+    do not delete an older instance that corresponds to this epoch.
+  */
+  atomic_uint_least64_t epoch;
+  /*
+    Number of tasks that this thread has running. Required so
+    that threads without tasks (and no longer are able to call
+    checkpoints) are not taken into account.
+  */
+  atomic_uint_least64_t nTasks;
+  /*
+    List of objects that are queued for deferred deletion, sorted by target_epoch. 
+    We defer deletion if we find that while scanning the list of all thread-specific
+    data, they are not up-to-date on that epoch. By sorting by target_epoch, we can
+    make shortcuts by removing a large bulk of data at once. The list is protected
+    by a lightweight spinlock, but is acquired using pseudo-TestAndSet operations
+    due to low contention.
+  */
+  atomic_uint_least64_t spinlock;
```

Comment:
does this spinlock protect deferList, recycleList, and next?

Reply:
`next` is atomic, `recycleList` is thread-local and should never be touched by another thread except during termination. The `spinlock` protects `deferList` as when a thread is about to be parked it will 'dump' its work on another active thread.

Reply:
then doesn't chpl_qsbr_defer_deletion_multi / chpl_qsbr_defer_deletion need to take the spinlock, since it calls push_defer_list?

Also get_recycle_list doesn't appear to be called anywhere, and pop_recycle_list has a comment saying it requires the spinlock.

Reply:
I tend to prefer to include that the lock is required in the function names, somewhere. Makes it easier to find oversights.